scipy/sparse/linalg/_eigen/_svds.py

```diff
@@ -341,46 +316,31 @@ def matmat_XH_X(x):
         eigvals, eigvec = eigsh(XH_X, k=k, tol=tol ** 2, maxiter=maxiter,
                                 ncv=ncv, which=which, v0=v0)
 
-    # Gramian matrices have real non-negative eigenvalues.
-    eigvals = np.maximum(eigvals.real, 0)
-
-    # Use the sophisticated detection of small eigenvalues from pinvh.
-    t = eigvec.dtype.char.lower()
-    factor = {'f': 1E3, 'd': 1E6}
-    cond = factor[t] * np.finfo(t).eps
-    cutoff = cond * np.max(eigvals)
-
-    # Get a mask indicating which eigenpairs are not degenerately tiny,
-    # and create the re-ordered array of thresholded singular values.
-    above_cutoff = (eigvals > cutoff)
-    nlarge = above_cutoff.sum()
-    nsmall = k - nlarge
-    slarge = np.sqrt(eigvals[above_cutoff])
-    s = np.zeros_like(eigvals)
-    s[:nlarge] = slarge
+    u = X_matmat(eigvec)
     if not return_singular_vectors:
-        return np.sort(s)
-
-    if n > m:
-        vlarge = eigvec[:, above_cutoff]
-        ularge = (X_matmat(vlarge) / slarge
-                  if return_singular_vectors != 'vh' else None)
-        vhlarge = _herm(vlarge)
+        s = svd(u, compute_uv=False)
+        return s[::-1]
+
+    # compute the right singular vectors of X and update the left ones accordingly
+    u, s, vh = svd(u, full_matrices=False)
+    u = u[:, ::-1]
+    s = s[::-1]
+    vh = vh[::-1]
+    return_u = (return_singular_vectors == 'u')
+    return_vh = (return_singular_vectors == 'vh')
```

Comment:
These two also should take into account that return_singular_values can also be True to set both to True. In hindsight this was a bad API design decision on our side.